### ISSUE TITLE
Firefox Extension?

### ISSUE THREAD

**Issue by TaylanTatli** [OllieTerrance/Homely#2]
Is it possible to make a firefox version? I don't like chrome but i'm using because your extension is so good.

**Comment by JourneyOver** [comment]
With the upcoming Mozilla WebExtension API coming soon that'll allow chrome extensions to work with firefox do you think you might have time to possibly look into doing this?

**Comment by OllieTerrance** [comment]
This is quite promising, though I can't find an equivalent to `chrome_url_overrides` in order to set the custom new tab page.  May just require some extra manifest stuff, or at least exposing a URL you can use as your home page.  In any case, this looks a lot more doable now!  😃

**Comment by JourneyOver** [comment]
Sounds great :D honestly been thinking about switching to firefox for a while now because chrome has been very buggy for me lately, but this is one addon/extension that is stopping me from actually doing so as every new tab page I've tried for firefox so far sucks.. So if you can find a way to do it, it would be great!

I did try setting things up with [Chrome Store Foxified](https://addons.mozilla.org/en-US/firefox/addon/chrome-store-foxified/) and a [tab switch overrider](https://addons.mozilla.org/en-US/firefox/addon/new-tab-override) on firefox and trying out chrome-extension://fplghnonomhbnlhdnmjpipoomkjdpeim/newtab.html but of course that method didn't work out sadly.

**Comment by OllieTerrance** [comment]
Looks like it's a bit broken as-is.  Steps to get it installed (using Firefox 49 developer):

1. Checkout a local copy of the repo.
2. Go to `about:debugging#addons`, check _Enable add-on debugging_.
3. Click _Load Temporary Add-on_, browse to and open **manifest.json** in the repo.
4. Homely should now appear in the extension list, click _Debug_ (might take a few seconds to load).
5. `chrome.extension.getURL("res/html/homely.html")` will give you the main page address.

But the stripy background is there, so that's something...  🎉

**Comment by JourneyOver** [comment]
Yep using chrome store foxified addon and going and installing the addon from the chrome store and mainly doing the same steps as above I was able to get the stripy background as well after it gave me the url of `moz-extension://ef82e03d-cf7f-4663-94dd-a676b07fb7c8/res/html/homely.html` so at least we can get to that point, now guess next thing is figuring out how to get everything else to work xD

**Comment by OllieTerrance** [comment]
Something quirky with the JavaScript going on, but no errors in the console...

More is happening than can be seen, because `body` is hidden until everything is ready.  If you remove the `display: none;` (line 5) from **res/css/homely.css**, we can see what's actually going on.

**Comment by OllieTerrance** [comment]
Ah, roadblock detected -- [`chrome.permissions` isn't supported yet](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/Chrome_incompatibilities#Planned_APIs).  Would mean having to declare all permissions upfront, which is not ideal given the scope, or drop things like notifications/baskets for now.

Also means diverging the code base to handle both types of permissions, which could get messy.  😒

**Comment by JourneyOver** [comment]
ooh wonder whats going on with the javascript to cause the issue with it not loading everything fully even though it's happening in the background.. and aww darn that sucks :( I didn't know the chrome.permissions wasn't supported yet, hopefully they get around to getting those supported soon enough.. 

Once they get them supported you wouldn't have to diverge the code base any right? so once it's supported you could theoretically possibly make this addon work with Firefox a whole lot more?

**Comment by OllieTerrance** [comment]
Hopefully, assuming it works the same as in Chrome, the two should behave the same.  I'll hold off officially porting it for now, but will keep an eye on what happens with the WebExtensions spec.

**Comment by JourneyOver** [comment]
Hopefully they will work the same as in chrome then, and alright that sounds fine :D glad that you are at least going to keep an eye on the WebExtensions stuff and hopefully be able to make an addon for firefox once development for it all opens up a bit more.

**Comment by OllieTerrance** [comment]
Quick update: looks like [optional permissions are scheduled for Firefox 54](https://wiki.mozilla.org/WebExtensions/Permissions#Optional_permissions_and_associated_permissions_API).

The Firefox nightlies have most of the features for required permissions (unsure about optional) -- I'll hopefully find some time to poke around with this soon.

**Comment by OllieTerrance** [comment]
Seeing as Chrome has been quite a pain lately, I've swapped back to Firefox for a bit.

54.0a2, no sign of `chrome.permissions` APIs yet.  But the only thing missing that breaks startup is `chrome.fontSettings`.  Removing the font iteration stuff will get you a working page, complete with block editing and saving.  🙂

![Homely on Firefox](https://cloud.githubusercontent.com/assets/4025899/24060980/89a1a214-0b4d-11e7-885c-b27b2d184d74.png)

```diff
diff --git a/Homely/res/js/homely.js b/Homely/res/js/homely.js
index e104a00..959aabf 100644
--- a/Homely/res/js/homely.js
+++ b/Homely/res/js/homely.js
@@ -2087,10 +2087,12 @@ $(document).ready(function() {
         }
         // request list of fonts from FontSettings API
-        chrome.fontSettings.getFontList(function fontsCallback(fonts) {
-            for (var i in fonts) {
-                $("#settings-style-font").append($("<option/>").text(fonts[i].displayName));
-            }
-            $("#settings-style-font").val(settings.style["font"]);
-        });
         $(".ext-name").text(manif.name);
         $(".ext-ver").text(manif.version);
```

Dev installation is cumbersome and doesn't persist across browser restarts, but for the curious:

* go to **about:debugging**
* click _Load Temporary Add-on_
* navigate to _Homely/manifest.json_
* once Homely appears in the list, click _Debug_, accept the remote debug prompt
* go to the _Inspect_ tab, content should have a body like this:
```html
<body>
  <script src="moz-extension://6f91c8f3-f622-44b6-a4e6-d491e2b54487/res/js/background.js"></script>
</body>
```
* go to **moz-extension://6f91c8f3-f622-44b6-a4e6-d491e2b54487/res/homely.html**, substituting for the UUID generated by your install

**Comment by JourneyOver** [comment]
Lookin good so far :D

**Comment by Terrance** [comment]
For those still lurking in here, the permissions API now exists, and the latest commits should now allow the extension to be installed in Firefox without any edits.

**Comment by antonsnotes** [comment]
@Terrance great! Does it mean that you are planning to add it to the Firefox Add-ons site?

**Comment by JourneyOver** [comment]
Yes please @Terrance can you get this up on the firefox add-ons site? so it's an easy click and install of the extension.
